### ISSUE TITLE
Deserialization issue

### ISSUE THREAD

**Issue by esso23** [wojtpl2/ExtendedXmlSerializer#156]
Hello there!
I'm having an issue trying to make this library to work and would appreciate some help.

I have an XML like this:
```xml
<TObjItemData>
      <Id>11546</Id>
      <Name>SomeObjectName</Name>
      ...
</TObjItemData>
```
it is a part of a larger XML, but i need just this one Element and I send it to the Deserialize method as text.
The class looks like this:

```cs
    [XmlRoot("TObjItemData")]
    public class DbD2Object : DataEntity
    {
        [XmlElement("Id")]
        public long Id { get; set; }

        [XmlElement("Name")]
        public string Name { get; set; }
        ....
    }
```

I tried all sorts of initialization of the serializer, but I think this should be the correct one (according to doc):
```cs
    IExtendedXmlSerializer serializer = new ConfigurationContainer()
        .ConfigureType<DbD2Object>().Name("TObjItemData")
        .Create();
```

Then I call the Deserialize<> method like this:

```cs
DbD2Object obj = serializer.Deserialize<DbD2Object>(element.ToString());
```

I tried all sorts of things but no matter what I do I always get the same exception.

```
System.InvalidOperationException: 'An attempt was made to load a type with the fully qualified name of 'TObjItemData', but no type could be located with that name.'
```

I honestly don't understand what "fully qualified name" means in this context, but I guess it has something to do with those xmlns attributes that are in every single XML in the documentation (never used them before tbh).
My problem is that I cannot change the input XML in any way, as I am receiving it from an external source and I'm using the library for deserialization only.

Thanks,
Esso

**Comment by Mike-EEE** [comment]
Hi @esso23 thanks for writing in.  The primary issue is that it appears that you are using the string from an `XmlElement` rather from a serialized object.  Try first creating your `DbD2Object` and call the serializer's `Serialize` method to create the string.  Calling `Deserialize` on that should get you the results that you are expecting.  Hopefully. 😄

**Comment by esso23** [comment]
Hi @Mike-EEE , thanks for the reply, but I think I don't understand. The XML itself is the serialized object and I need to deserialize the XML into a C# object. I don't create `DbD2Object` in my code at any point, I need it to be created from the received XML using deserialization. I cannot call the serialize method as there is nothing to serialize. Am I missing something?

**Comment by Mike-EEE** [comment]
Ah, OK.   Thanks for the information @esso23.

FWIW the typical expected workflow is serialize the object graph with the serializer, storing (or sending it) and then deserializing it with the same type of serializer.  When you say that you are receiving it, how is it being sent?  Or rather, how is it being serialized then sent? :)

Are you able to serialize/deserialize this XML with the classic `XmlSerializer`?  If so, please post the XML here (along with any code used to configure the `XmlSerializer`) and I will take a look at it.  It sounds like the type names are not being mapped properly.

**Comment by esso23** [comment]
I am receiving this XML from an external source so I have no control over the serialization of the XML. It's just some 3rd party application.

I am able to serialize and deserialize using standard XmlSerializer.
```
            XmlSerializer serializer = new XmlSerializer(typeof(DbD2Object))
            DbD2Object obj = (DbD2Object)serializer.Deserialize(element.CreateReader());
```

The reason why I want to use this library is that the standard serializer cannot handle the DateTimes in XML, so the deserialization fails at that point. I may also need some other ways of customization this library provides in the future.

XML (this is the exact content of element.ToString() from the snippet I posted earlier)
```xml
    <TObjItemData>
      <Id>11546</Id>
      <Name>name</Name>
      <Descript>descript</Descript>
      <Typ>POINT</Typ>
      <ParentId>BIW_L</ParentId>
      <Value_Type>Ci</Value_Type>
      <CreateTime>06.11.2017 00:07:39.541</CreateTime>
      <ModifyTime>06.11.2017 00:07:39.541</ModifyTime>
      <Text_Palette/>
      <Read_Only>False</Read_Only>
      <TECH_UNIT/>
      <UseLimits>False</UseLimits>
      <VHL>1.00000000000000E+38</VHL>
      <HL>1.00000000000000E+38</HL>
      <LL>-1.00000000000000E+38</LL>
      <VLL>-1.00000000000000E+38</VLL>
      <VHL_ID/>
      <HL_ID/>
      <LL_ID/>
      <VLL_ID/>
      <Hysterezia>0.00000000000000E+00</Hysterezia>
      <TransPalIdx>0</TransPalIdx>
      <Hidden>False</Hidden>
      <uuid>BDCCD35F6932404A8BEE1C35500581EF</uuid>
      <memberOfModule/>
      <moduleRelation>AM_None</moduleRelation>
    </TObjItemData>
```

The class (you can ignore the DBField attributes)
```cs
    [XmlRoot("TObjItemData")]
    [DbTable("objlist")]
    public class DbD2Object : DataEntity
    {
        [XmlElement("Id")]
        [IsIdField]
        [DbField("id")]
        public long Id { get; set; }

        [XmlElement("Name")]
        [DbField("name")]
        public string Name { get; set; }

        [XmlElement("Descript")]
        [DbField("descript")]
        public string Descript { get; set; }

        [XmlElement("Typ")]
        [DbField("typ")]
        public D2ObjectType ObjectType { get; set; }

        [XmlIgnore]
        [DbField("parent")]
        public long ParentId { get; set; }

        [XmlElement("Value_Type")]
        [DbField("value_type")]
        public D2ValueType ValueType { get; set; }

        [XmlIgnore]
        [DbField("refnr")]
        public short RefNr { get; set; }

        [XmlElement("ModifyTime", DataType = "date")]
        [DbField("modify_time")]
        public DateTime ModifyTime { get; set; }

        [XmlIgnore]
        [DbField("text_palette")]
        public long TextPaletteId { get; set; }

        [XmlElement("TECH_UNIT")]
        [DbField("tech_unit")]
        public string TechUnit { get; set; }

        [XmlElement("VHL")]
        [DbField("vhl")]
        public double VHL { get; set; }

        [XmlElement("HL")]
        [DbField("hl")]
        public double HL { get; set; }

        [XmlElement("LL")]
        [DbField("ll")]
        public double LL { get; set; }

        [XmlElement("VLL")]
        [DbField("vll")]
        public double VLL { get; set; }

        [XmlIgnore]
        [DbField("vhl_id")]
        public long VHLid { get; set; }

        [XmlIgnore]
        [DbField("hl_id")]
        public long HLid { get; set; }

        [XmlIgnore]
        [DbField("ll_id")]
        public long LLid { get; set; }

        [XmlIgnore]
        [DbField("vll_id")]
        public long VLLid { get; set; }

        [XmlIgnore]
        [DbField("pasrefnr")]
        public int PasRefNr { get; set; }

        [XmlElement("Read_Only")]
        [DbField("read_only", true)]
        public bool ReadOnly { get; set; }

        [XmlElement("Hysterezia")]
        [DbField("hysterezia")]
        public double Hysterezia { get; set; }

        [XmlIgnore]
        [DbField("vhl_row")]
        public short VHLRow { get; set; }

        [XmlIgnore]
        [DbField("vhl_col")]
        public short VHLCol { get; set; }

        [XmlIgnore]
        [DbField("hl_row")]
        public short HLRow { get; set; }

        [XmlIgnore]
        [DbField("hl_col")]
        public short HLCol { get; set; }

        [XmlIgnore]
        [DbField("ll_row")]
        public short LLRow { get; set; }

        [XmlIgnore]
        [DbField("ll_col")]
        public short LLCol { get; set; }

        [XmlIgnore]
        [DbField("vll_row")]
        public short VLLRow { get; set; }

        [XmlIgnore]
        [DbField("vll_col")]
        public short VLLCol { get; set; }

        [XmlElement("TransPalIdx")]
        [DbField("trans_pal_idx")]
        public int TransPalIdx { get; set; }

        [XmlElement("Hidden")]
        [DbField("hidden_object", true)]
        public bool HiddenObject { get; set; }

        [XmlElement("uuid")]
        [DbField("uuid")]
        public string UUID { get; set; }

        [XmlIgnore]
        [DbField("version_of_uuid")]
        public string VersionOfUUID { get; set; }

        [XmlElement("memberOfModule")]
        [DbField("memberofmodule")]
        public long MemberOfModule { get; set; }

        [XmlElement("moduleRelation")]
        [DbField("modulerelation")]
        public short ModuleRelation { get; set; }

        [XmlElement("CreateTime", DataType = "date")]
        [DbField("create_time")]
        public DateTime CreateTime { get; set; }

        //-------------------------------------

        
        [XmlElement("ParentId")]
        public string ParentName
        {
            get => mParentName;
            set
            {
                mParentName = value;
                ParentId = UpdateId(value);
            }
        }
        private string mParentName;

        [XmlElement("Text_Palette")]
        public string TextPaletteName
        {
            get => mTextPaletteName;
            set
            {
                mTextPaletteName = value;
                TextPaletteId = UpdateId(value);
            }
        }
        private string mTextPaletteName;

        [XmlElement("VHL_ID")]
        public string VHLidName
        {
            get => mVHLidName;
            set
            {
                mVHLidName = value;
                VHLid = UpdateId(value);
            }
        }
        private string mVHLidName;

        [XmlElement("HL_ID")]
        public string HLidName
        {
            get => mHLidName;
            set
            {
                mHLidName = value;
                HLid = UpdateId(value);
            }
        }
        private string mHLidName;

        [XmlElement("LL_ID")]
        public string LLidName
        {
            get => mLLidName;
            set
            {
                mLLidName = value;
                LLid = UpdateId(value);
            }
        }
        private string mLLidName;

        [XmlElement("VLL_ID")]
        public string VLLidName
        {
            get => mVLLidName;
            set
            {
                mVLLidName = value;
                VLLid = UpdateId(value);
            }
        }
        private string mVLLidName;

        // Some methods
        // ...
    }
```

**Comment by Mike-EEE** [comment]
This is actually a good point here.  We do offer implicit typing:

https://github.com/wojtpl2/ExtendedXmlSerializer#implicit-namespacestyping

But we do not apply it unless it's explicit.  What we could do for the `Serialize`/`Deserialize` methods is include that for the type that is being requested.  So for this scenario it *should* deserialize just fine.

Try modifying your configuration container to call `EnableImplicitTyping(typeof(DbD2Object))` and see if that gets you further along, @esso23.  I'm actually really interested in seeing if this works because that is one wicked DTO graph. 😆

**Comment by esso23** [comment]
Well, now the deserialization works (which makes me really happy :D), but it fails with the exact same exception as the standard XmlSerializer:
```
System.FormatException: 'The string '06.11.2017 00:07:39.541' is not a valid AllXsd value.'
```
So I created a IConverter<> like this:
```cs
        private class DateTimeConverter : IConverter<DateTime>
        {
            public static readonly DateTimeConverter Default = new DateTimeConverter();

            public bool IsSatisfiedBy(TypeInfo parameter)
            {
                if (parameter.AsType() == typeof(DateTime))
                    return true;
                return false;
            }

            public DateTime Parse(string data)
            {
                return DateTime.ParseExact(data, "dd.MM.yyyy HH:mm:ss.fff", CultureInfo.InvariantCulture);
            }

            public string Format(DateTime instance)
            {
                return instance.ToString("dd.MM.yyyy HH:mm:ss.fff");
            }
        }
```
And my registration looks like this:
```cs
    IExtendedXmlSerializer serializer = new ConfigurationContainer()
                .EnableImplicitTyping(typeof(DbD2Object))
                .Register(DateTimeConverter.Default)
                .Create();
```

But the converter is only called for the `DbD2Object` itself and for enums. Is there any way how to make the library call it for deserialization of DateTime aswell? Or am I doing it completely wrong?

**Comment by Mike-EEE** [comment]
Boo... this just got complicated, LOL.  The reason why your converter isn't working is that there is already a converter that is registered via the `WellKnownConverters` class that handles the `DateTime` type.  We only expose `Register` and `Unregister`, assuming that you are using custom types that are not well-known.  We are obviously in error with this thinking. 😉 

What we need to do is add a `Find<>` which would find the first registered converter that matches the type, and then you could remove it and register a new one.  Or maybe better yet, have the `Register` method remove all registered converters that match the passed-in type before adding the new one. That way there is only one converter per-type.

Unfortunately I do not have a good answer here to unblock you.  Are you able to add this project to your solution as a CSPROJ reference to get you where you would like to be?  You could then modify the code to get it how you would like if you feel so brave. :)

The other option is to use XSLT to format your data so that it is in the expected format before passing it into a serializer. :)

**Comment by esso23** [comment]
Hello @Mike-EEE. Thanks for the advice, I managed to make it work. I based my solution on your suggestion and the code now keeps only one converter per type.
I gotta be honest though. Your code is kinda complicated (for me) and I'm not very confident about making a pull request.
Do you have any plans on implementing this in near future?

```diff
diff --git a/src/ExtendedXmlSerializer/ExtensionModel/Content/Extensions.cs b/src/ExtendedXmlSerializer/ExtensionModel/Content/Extensions.cs
index 2f0ce05..60691a4 100644
--- a/src/ExtendedXmlSerializer/ExtensionModel/Content/Extensions.cs
+++ b/src/ExtendedXmlSerializer/ExtensionModel/Content/Extensions.cs
@@ -29,6 +29,7 @@ using ExtendedXmlSerializer.Core.Sources;
 using ExtendedXmlSerializer.Core.Specifications;
 using ExtendedXmlSerializer.ExtensionModel.Content.Members;
 using System;
+using System.Linq;
 
 namespace ExtendedXmlSerializer.ExtensionModel.Content
 {
@@ -113,14 +114,23 @@ namespace ExtendedXmlSerializer.ExtensionModel.Content
 
 		public static IConfigurationContainer Register<T>(this IConfigurationContainer @this, IConverter<T> converter)
 		{
-			@this.Root.Find<ConverterRegistryExtension>()
-			     .Converters.Add(Converters<T>.Default.Get(converter));
+			ICollection<IConverter> converters = @this.Root.Find<ConverterRegistryExtension>().Converters;
+
+			IConverter converterToRemove = converters.FirstOrDefault(c => c.Get().AsType() == typeof(T));
+			if (converterToRemove != null)
+				@this.Unregister(converterToRemove);
+
+			converters.Add(Converters<T>.Default.Get(converter));
 			return @this;
 		}
 
 		public static bool Unregister<T>(this IConfigurationContainer @this, IConverter<T> converter)
 			=> @this.Root.Find<ConverterRegistryExtension>()
-			        .Converters.Remove(Converters<T>.Default.Get(converter));
+					.Converters.Remove(Converters<T>.Default.Get(converter));
+
+		public static bool Unregister(this IConfigurationContainer @this, IConverter converter)
+			=> @this.Root.Find<ConverterRegistryExtension>()
+				.Converters.Remove(converter);
 
 		sealed class Converters<T> : ReferenceCache<IConverter<T>, IConverter>
 		{
``

**Comment by Mike-EEE** [comment]
Coooool thanks for the update @esso23!  Yeah I will be the first to admit our fluent api/configuration integration is a mess.  It really needs a do over from scratch, something we have on our minds for v3.  The other consideration is that we wanted to keep everything as internal as possible for v2 and expose it as we need it as we went along.  

I think in this case what we should do is expose all the default extensions used in `DefaultExtensions` so that extenders can easily add their own set of extensions if they would like, especially in scenarios such as this where you could have easily replaced the `DefaultExtensions` class and point it to a new list of extensions, one that would instantiate a new `ConverterRegistryExtension` with your new converter in that list.

What do you think about that, @wojtpl2?  Are you OK with exposing these classes?

As for your fix here @esso23, I am in the middle of some refactoring for #150 and some of the classes touch this use case.  I will see if it's in a state to commit and if so try to get this addresses sometime today for you.

**Comment by esso23** [comment]
Thanks for the reply @Mike-EEE .
Giving me the tools to micro-manage things on the inside would be amazing. Imho, people who choose to use this library instead of standard XmlSerializer want exactly that - because with default XmlSerializer you either use what it gives you or you write the whole code for serialization/deserialization yourself.
Solution you propose sounds great!

**Comment by Mike-EEE** [comment]
Yeah technically there isn't anything stopping you from simply copying/pasting the code as it is now into your own namespace and changing the code you need, but that is a hassle.  I can understand the need to keep things internal (especially at the start), but ideally you do want to expose components over time.  The classic `XmlSerializer` has a lot to account for so it's not so easy for them to flip a switch, not to mention it was made in the time when closed-source was the king. 😉 

In any case, I think you're right here.  I don't think @wojtpl2 will have a problem with it either.  I will go ahead and get this checked in and we can always make more commits to reverse any issues that are discovered.  As long as we do it before a major (or minor in this case: 2.1) release then we're OK.  Generally speaking, heh heh.

**Comment by Mike-EEE** [comment]
Alright @esso23 you should be set to go here.  I've updated the converter registrations to utilize a component we already had within our code, which was taken from the classic .NET Framework (`KeyedByTypeCollection`) and ensures that items that are added are keyed by their type and only get added once.  This should throw any errors in case there are two objects with the same type, but it also removes items as well before adding it.

Additionally, I've exposed all components around the `DefaultExtensions` and you should be able to extend extensions more easily now.

You can find the latest version here on our preview feed:
https://www.myget.org/F/wojtpl2/api/v3/index.json

**Comment by Mike-EEE** [comment]
Going to close this issue now.  Please let me know if there are any further problems you find with it and we will reopen it.  Please also feel free to open a new issue altogether if you feel that is more appropriate. 👍
